### PR TITLE
feat(returns): add rustledger-returns crate with XIRR (money-weighted return)

### PR DESCRIPTION
**Pull request by robcohen**
## What

The first slice of investment-returns reporting (#1814): a new `rustledger-returns` leaf crate holding the **money-weighted return (XIRR)** math.

Per the plan in #1814, the returns math is a **pure engine crate** so every eventual consumer (CLI `report returns`, BQL, the FFI component, rustfava) reuses one implementation rather than re-deriving it — canonical-function discipline. This PR is deliberately just the numeric core: no ledger loading, no price DB, no I/O.

- `CashFlow { date, amount }` with an **investor-centric** sign convention (money put in = negative; sale proceeds, dividends, and the terminal market value of the held position = positive).
- `xirr(&[CashFlow]) -> Option<f64>` — annualized IRR over an irregularly-spaced series via Newton's method with a bisection fallback; actual/365 day count (matching spreadsheet `XIRR` and beangrow). Returns `None` when there's no sign change, fewer than two flows, or a non-finite result.

## Why this scope

XIRR is the novel, self-contained, provably-correct-in-isolation part — the right thing to land and de-risk first. The follow-ups build on it:
- **Cash-flow extraction** (group boundary + the internal-transfer exclusion, dividend classification, currency conversion, terminal valuation) — this is where correctness lives, and where `PriceDatabase` gets relocated into `rustledger-core`.
- **Time-weighted return** (Modified Dietz / true TWR) — needs the per-date portfolio valuations only the extraction layer can supply, so it lands with it.
- **`rledger report returns --investments <pattern>`**, then the BQL/FFI surfaces.

## Tests

`cargo test -p rustledger-returns` (7, all pass):
- the canonical spreadsheet XIRR example → ≈ 0.3734;
- a clean 10% one-year case and a −10% case (exact-ish asserts);
- solver agreement (NPV ≈ 0 at the returned rate on a 4-flow series);
- `None` for no-sign-change, <2 flows, and a zero-flow-doesn't-count case.

Clippy/doc/fmt clean. Crate registered in the workspace members/default-members and the crates.io publish `CRATES` list (sync check passes).

Part of #1814.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_013gJH7uBBF9Xw9et9QFRcsU

### CODE REVIEW

**Comment by robcohen** [comment]
**Deep review** (16 agents, adversarial numerical verification): 5 real correctness defects in the solver, all fixed in 025a85ee. This is exactly the kind of math where adversarial review pays off — the happy-path tests were green but the solver returned wrong or fabricated rates on several inputs:

1. **Absolute NPV tolerance** (1e-7) was unreachable for large portfolios and trivially satisfied for tiny ones (any rate accepted as a root). Now a **relative tolerance** scaled to gross flow magnitude.
2. **Same-date flows** made NPV rate-independent → Newton's first check passed at the 0.1 seed → xirr **fabricated a 10% return**. Now a degenerate guard returns `None`.
3. **Newton's stall branch** returned the stalled rate without checking it was a root — a step underflowing near the rate=-1 pole surfaced as a fake IRR and suppressed the bisection fallback. Now it only returns on a genuine NPV check.
4. **Bisection lower bound** (-0.9999) couldn't reach IRRs below -99.99% (legitimate near-total losses). Lowered to -0.9999999 with explicit endpoint-root checks.
5. **Bisection `f_a * f_mid < 0`** could overflow to ±inf for large NPV magnitudes and misclassify the sub-interval. Replaced with a sign comparison.

Plus: npv + derivative share one pass (halves `powf` per Newton step).

Four new edge-case tests pin the fixed paths: same-date → None, near-total-loss (~-99.995%), scale invariance across 1e6/1e-6 magnitudes, and a multiple-sign-change series asserting the returned rate is a genuine root. The canonical spreadsheet XIRR reference (~0.3734) still holds. 11 tests, clippy/doc/fmt clean.

File: crates/rustledger-returns/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    let series: Vec<(f64, f64)> = flows
+        .iter()
+        .map(|f| {
+            let days = f.date.since(origin).map_or(0, |span| span.get_days());
+            (
+                f64::from(days) / DAYS_PER_YEAR,
+                f.amount.to_f64().unwrap_or(0.0),
+            )
+        })
+        .collect();
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger-returns/src/lib.rs]: Fixed — the series build now collects into Option and propagates None, so an unrepresentable Decimal yields no rate instead of a silently-zeroed flow. (In practice rust_decimal always fits in f64 range, but the defensive path removes the footgun.)

**Comment by robcohen** [comment]
**Round-2 deep review** (16 agents): found that round-1's bound-lowering was incomplete — it could still return a garbage rate near the rate=-1 pole. Fixed in 1acfafb8, but with a better lever than more bound-tuning:

- **Final root verification.** `xirr` now checks the solved candidate is a genuine root (`|NPV| <= tol`) before returning it. Any path that ended on an iteration limit, a stalled Newton step, or a bracket corrupted by a non-finite NPV near the pole (a far-dated flow whose discount factor overflows) now yields `None` — never a fabricated/garbage rate. Genuine results sit well within tol, so valid roots are never rejected. This single guard neutralizes the *dangerous* half of every pole/NaN finding.
- **Documented the true limits** instead of chasing them (lowering the bracket bound worsens the overflow): IRRs at/below ~-99.99999%/yr (near-total loss, against the pole) return `None` gracefully; even-multiplicity roots (NPV touches zero without crossing) may not bracket.

Also fixed a **vacuous test** the review flagged: `scale_invariance`'s tiny case resolved under the 1e-12 floor with the seed already a root, so it never exercised relative tolerance. Replaced with `tiny_magnitude_requires_relative_tolerance` (~1e-8 flows, ~50% return — where an absolute 1e-7 tolerance *would* accept the 0.10 seed as a bogus root, so it genuinely distinguishes relative from absolute). Added `extreme_loss_below_supported_range_returns_none` pinning graceful-None.

13 tests, clippy/doc/fmt clean.

**Comment by robcohen** [comment]
**Round-3 deep review** (Brent-focused): found that my round-2 final guard was itself a bug — and it contradicted my own pre-review analysis, which is precisely why the independent pass earns its keep.

**The defect:** near the rate=-1 pole the NPV derivative is enormous, so a rate Brent pins to full f64 precision can still have `|NPV|` a hair above `tol`. My round-2 `|NPV| <= tol` re-check then **wrongly rejected that genuine root** and returned `None` for a legitimately-bracketed deep loss (verified: `-0.9999995`, inside the bracket, returned None). I'd reasoned the `fb<tol` exit would fire first; simulation proved it doesn't.

**The fix** — recognize that a **converged, finite, sign-changing bracket is an IVT proof of a root** regardless of NPV magnitude:
- Dropped xirr's `|NPV| <= tol` re-check. Both solvers already self-verify (Newton re-checks NPV at a stalled step; Brent returns only on genuine convergence), so no non-root can reach the caller.
- Closed the *actual* garbage path at its source: Brent now rejects a bracket whose endpoint NPV is non-finite (a far-dated flow overflowing against the pole), so a corrupted bracket never reaches refinement.
- Brent's iteration-limit tail returns `None` (non-convergence), not an unconverged `b`.

**Second finding (test observability, per canonical-function discipline):** no test observed *which* internal Brent branch ran, so a corrupted inverse-quadratic formula could ship masked by the bisection safeguard. Added a test-only IQI-accepted counter + `brent_exercises_inverse_quadratic_interpolation` asserting the IQI branch actually runs. Plus `deep_loss_within_bracket_is_found` pinning the fixed behavior (was None, now the rate).

16 tests, clippy/doc/fmt clean.
